### PR TITLE
Feature/wjiang/#149 replace set eta

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
Data for ak, bk are set in the source code m_set_eta.F90. If users want their own set of data, they have to change the source code. Worse, if the number of levels is same but the actual data is different from that in the source code. A new MAPL_VerticalGrid.F90 is added to solve this issue.  The users can provide they own eta data through config (or RC)  file
## Description

The module provide get_eta subroutine to read data from config file . get_pressure_levels is also provided
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [ ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [ ] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [ ] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: MAPL_Base/CMakeLists.txt [hunk 1]
```diff
@@ -26,7 +26,7 @@ set (srcs
   MAPL_Cap.F90         hash.c
   MAPL_CapGridComp.F90
   MAPL_GridType.F90 
-  MAPL_GridSpec.F90        MAPL_VerticalMethods.F90 MAPL_newCFIOitem.F90
+  MAPL_GridSpec.F90        MAPL_VerticalMethods.F90 MAPL_newCFIOitem.F90 MAPL_EtaHybridVerticalCoordinate.F90
```
Comment (tclune) [inline MAPL_Base/CMakeLists.txt]: Split this line

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 [hunk 2]
```diff
@@ -0,0 +1,282 @@
+#include "MAPL_Generic.h"
+#include "MAPL_ErrLog.h"
+#include "unused_dummy.H"
```
Comment (tclune) [inline MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: I think only the middle #include is needed here.

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      integer, optional, intent(inout) :: rc
+      real(kind=REAL64), allocatable :: ak(:)
+      real(kind=REAL64), allocatable :: bk(:)
+      integer :: status
+      integer :: k, num_levels
+      real(kind=REAL64) :: ptop, pint, ref_pressure
+      character(len=32) :: data_label
+ 
+      call ESMF_ConfigGetAttribute(config, num_levels,label='NUM_LEVELS:', __RC__ )
+      call ESMF_ConfigGetAttribute(config, ref_pressure,label='REF_PRESSURE:', default = DEFAULT_REFERENCE_PRESSURE, __RC__ )
```
Comment (tclune) [inline MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: split line

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+      procedure :: get_pressures_r8_3d
+      procedure :: get_pressures_r4_3d
+      generic :: get_eta =>get_eta_r8, get_eta_r4
+      generic :: get_pressure_levels=>get_pressure_levels_r8,   get_pressure_levels_r4
+      generic :: get_pressures      =>get_pressures_r8_3d,get_pressures_r4_3d
+   end type EtaHybridVerticalCoordinate
+
+   interface newEtaHybridVerticalCoordinate
+      module procedure new_EtaHybridVerticalCoordinate_by_ak_bk
+      module procedure new_EtaHybridVerticalCoordinate_by_cfg
```
Comment (tclune) [inline MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: Need one more constructor where just the filename is passed in.   Then a config file is created and the existing constructor is called.

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      integer :: num_levels = 0
+      real(kind=REAL64) :: ref_pressure
+   contains
+      procedure :: get_eta_r8
+      procedure :: get_eta_r4
+      procedure :: get_pressure_levels_r8
+      procedure :: get_pressure_levels_r4
+      procedure :: get_pressures_r8_3d
+      procedure :: get_pressures_r4_3d
+      generic :: get_eta =>get_eta_r8, get_eta_r4
```
Comment (tclune) [inline MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: Add a module level interface that creates the object and returns ak, bk all in one swoop.

```f90
interface get_eta
   module procedure get_eta_onestep_r32
   module procedure get_eta_onestep_r64
end interface
```

```f90
subroutine get_eta_onstep_r32(filename, ptop, pint, ak, bk, unused, rc)
    ...
     type (EtaHybridVerticalCoordinate) :: vgrid
     vgrid = EtaHybridVerticalCoordinate(filename)
    call vgrid%get_eta(...)
end subroutine
```

File: MAPL_Base/CMakeLists.txt (hunk 1 above)
Comment (weiyuan-jiang) [reply MAPL_Base/CMakeLists.txt]: done

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 (hunk 2 above)
Comment (weiyuan-jiang) [reply MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: we only need the first one

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 (hunk 3 above)
Comment (weiyuan-jiang) [reply MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: done

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 (hunk 4 above)
Comment (weiyuan-jiang) [reply MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: done

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 (hunk 5 above)
Comment (weiyuan-jiang) [reply MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: done. but 1) do wee need "get_pressure" interface? 2) do we need to make EtaHybridVerticalCoordinate private?

File: MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90 [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #include "MAPL_Generic.h"
 
-module MAPL_EtaHybridVerticalCoordinate
+module MAPL_EtaHybridVerticalCoordinateMod
```
Comment (tclune) [inline MAPL_Base/MAPL_EtaHybridVerticalCoordinate.F90]: Note that the _Mod suffix will be going away in newer versions of MAPL.   But not worth fixing back now.

**Comment by weiyuan-jiang** [comment]
@mathomp4 Why the checks failed? How can see the detail failed message?

**Pull request merged**